src/metabase/db/liquibase.clj

```diff
@@ -116,6 +116,35 @@
     (liquibase.h2/h2-database liquibase-conn)
     (.findCorrectDatabaseImplementation (DatabaseFactory/getInstance) liquibase-conn)))
 
+(when config/is-test?
+  ;; A lot of time during migration testing is spent parsing Liquibase migration YAMLs. The code below adds cache
+  ;; for the parsed files by implementing a custom caching parser.
+  (def ^:private ^java.util.concurrent.ConcurrentHashMap -test-only-migration-file-cache
+    (java.util.concurrent.ConcurrentHashMap.))
+  (def ^:private ^:dynamic -test-only-*return-caching-parser* true)
+
+  (def ^:private -test-only-caching-parser
+    (reify liquibase.parser.ChangeLogParser
+      (getPriority [_] 1024) ;; So that it has the highest priority in the parser factory.
+      (parse [_ physicalChangeLogLocation changeLogParameters resourceAccessor]
+        (let [key [physicalChangeLogLocation (.getDatabase changeLogParameters)]]
+          (or (.get -test-only-migration-file-cache key)
+              ;; Need to rebind the dynvar so that the ParserFactory doesn't return the caching parser again.
+              (let [real-parser (binding [-test-only-*return-caching-parser* false]
+                                  (.getParser (liquibase.parser.ChangeLogParserFactory/getInstance)
+                                              physicalChangeLogLocation resourceAccessor))
+                    parsed (.parse ^liquibase.parser.ChangeLogParser real-parser
+                                   physicalChangeLogLocation changeLogParameters resourceAccessor)]
+                (.put -test-only-migration-file-cache key parsed)
```

Comment:
Rather than a `.get` and a `.put`, why not use the atomic `.computeIfAbsent` instead?

Reply:
It is because ConcurrentHashMap doesn't allow recursive modifications (if something tries to write into it inside `computeIfAbsent`.

Reply:
Ah ok, I didn't realize the parser could recurse like that.